server/channels/db/migrations/postgres/000107_threadmemberships_cleanup.up.sql

```diff
@@ -0,0 +1,12 @@
+DELETE FROM threadmemberships WHERE (postid, userid) IN (
+    SELECT
+        threadmemberships.postid,
+        threadmemberships.userid
+    FROM
+        threadmemberships
+        JOIN threads ON threads.postid = threadmemberships.postid
+        LEFT JOIN channelmembers ON channelmembers.userid = threadmemberships.userid
+            AND threads.channelid = channelmembers.channelid
+    WHERE
+        channelmembers.channelid IS NULL
+);
```

Comment:
I tested this on my 12M database in Postgres:
```suggestion
DELETE FROM threadmemberships WHERE (postid, userid) IN (
    SELECT
        threadmemberships.postid,
        threadmemberships.userid
    FROM
        threadmemberships
        JOIN threads ON threads.postid = threadmemberships.postid
    WHERE
        (threadmemberships.UserId, threads.ChannelId) NOT IN (SELECT UserId, ChannelId FROM channelmembers)
);
```
The explain analyze was taking minutes and had to cancel it, so the original query is good.

Reply:
I thought I had executed it, but maybe I hadn't, because I did when you asked this, forgot about it, and just recalled it was running:
```sql
> psql -U mmuser mattermost
psql (15.2 (Ubuntu 15.2-1.pgdg22.04+1), server 12.14 (Ubuntu 12.14-1.pgdg22.04+1))
Type "help" for help.

mattermost=> begin;
BEGIN
mattermost=*> explain analyze DELETE FROM threadmemberships WHERE (postid, userid) IN (
    SELECT
        threadmemberships.postid,
        threadmemberships.userid
    FROM
        threadmemberships
        JOIN threads ON threads.postid = threadmemberships.postid
        LEFT JOIN channelmembers ON channelmembers.userid = threadmemberships.userid
            AND threads.channelid = channelmembers.channelid
    WHERE
        channelmembers.channelid IS NULL
);

                                                                                      QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Delete on threadmemberships  (cost=2273421.65..2512202.14 rows=3532236 width=24) (actual time=8087878.620..8087878.625 rows=0 loops=1)
   ->  Hash Semi Join  (cost=2273421.65..2512202.14 rows=3532236 width=24) (actual time=21133.230..23666.475 rows=2766690 loops=1)
         Hash Cond: (((threadmemberships.postid)::text = (threads.postid)::text) AND ((threadmemberships.userid)::text = (threadmemberships_1.userid)::text))
         ->  Seq Scan on threadmemberships  (cost=0.00..82419.36 rows=3532236 width=60) (actual time=213.854..496.287 rows=3532236 loops=1)
         ->  Hash  (cost=2236850.82..2236850.82 rows=1194122 width=99) (actual time=20914.581..20914.585 rows=2766690 loops=1)
               Buckets: 32768 (originally 32768)  Batches: 128 (originally 64)  Memory Usage: 3841kB
               ->  Merge Anti Join  (cost=2174127.37..2236850.82 rows=1194122 width=99) (actual time=16056.838..20170.902 rows=2766690 loops=1)
                     Merge Cond: (((threadmemberships_1.userid)::text = (channelmembers.userid)::text) AND ((threads.channelid)::text = (channelmembers.channelid)::text))
                     ->  Sort  (cost=1101804.95..1110635.54 rows=3532236 width=120) (actual time=8964.121..10666.647 rows=3532236 loops=1)
                           Sort Key: threadmemberships_1.userid, threads.channelid
                           Sort Method: external merge  Disk: 449368kB
                           ->  Hash Join  (cost=106322.79..282999.31 rows=3532236 width=120) (actual time=315.090..1937.437 rows=3532236 loops=1)
                                 Hash Cond: ((threadmemberships_1.postid)::text = (threads.postid)::text)
                                 ->  Seq Scan on threadmemberships threadmemberships_1  (cost=0.00..82419.36 rows=3532236 width=60) (actual time=0.011..383.017 rows=3532236 loops=1)
                                 ->  Hash  (cost=86640.35..86640.35 rows=846835 width=60) (actual time=314.924..314.925 rows=847646 loops=1)
                                       Buckets: 65536  Batches: 32  Memory Usage: 2897kB
                                       ->  Seq Scan on threads  (cost=0.00..86640.35 rows=846835 width=60) (actual time=0.018..155.678 rows=847646 loops=1)
                     ->  Sort  (cost=1072322.18..1084012.75 rows=4676228 width=60) (actual time=7092.653..8467.265 rows=4676227 loops=1)
                           Sort Key: channelmembers.userid, channelmembers.channelid
                           Sort Method: external merge  Disk: 320360kB
                           ->  Seq Scan on channelmembers  (cost=0.00..202637.28 rows=4676228 width=60) (actual time=0.019..486.143 rows=4676228 loops=1)
 Planning Time: 3.440 ms
 JIT:
   Functions: 33
   Options: Inlining true, Optimization true, Expressions true, Deforming true
   Timing: Generation 0.742 ms, Inlining 30.721 ms, Optimization 104.356 ms, Emission 78.847 ms, Total 214.666 ms
 Execution Time: 8087964.315 ms
(27 rows)
```

It took more than two hours :thinking: I'll run the modified one now, since I didn't wait more than 5 minutes when I tried.

Reply:
Wait, I think that's wrong, scratch that, I don't know if I was running something else at the same time and got a lock? I'll repeat these two later or tomorrow and come back with actual times.

Reply:
Now this makes more sense. The original query took ~30 seconds:
```
mattermost=*> EXPLAIN ANALYZE DELETE FROM threadmemberships WHERE (postid, userid) IN (
    SELECT
        threadmemberships.postid,
        threadmemberships.userid
    FROM
        threadmemberships
        JOIN threads ON threads.postid = threadmemberships.postid
        LEFT JOIN channelmembers ON channelmembers.userid = threadmemberships.userid
            AND threads.channelid = channelmembers.channelid
    WHERE
        channelmembers.channelid IS NULL
);
                                                                                      QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Delete on threadmemberships  (cost=2273421.65..2512202.14 rows=3532236 width=24) (actual time=30746.086..30746.092 rows=0 loops=1)
   ->  Hash Semi Join  (cost=2273421.65..2512202.14 rows=3532236 width=24) (actual time=20018.134..22436.471 rows=2766690 loops=1)
         Hash Cond: (((threadmemberships.postid)::text = (threads.postid)::text) AND ((threadmemberships.userid)::text = (threadmemberships_1.userid)::text))
         ->  Seq Scan on threadmemberships  (cost=0.00..82419.36 rows=3532236 width=60) (actual time=233.870..550.646 rows=3532236 loops=1)
         ->  Hash  (cost=2236850.82..2236850.82 rows=1194122 width=99) (actual time=19783.765..19783.770 rows=2766690 loops=1)
               Buckets: 32768 (originally 32768)  Batches: 128 (originally 64)  Memory Usage: 3841kB
               ->  Merge Anti Join  (cost=2174127.37..2236850.82 rows=1194122 width=99) (actual time=15370.599..19125.528 rows=2766690 loops=1)
                     Merge Cond: (((threadmemberships_1.userid)::text = (channelmembers.userid)::text) AND ((threads.channelid)::text = (channelmembers.channelid)::text))
                     ->  Sort  (cost=1101804.95..1110635.54 rows=3532236 width=120) (actual time=8119.338..9697.393 rows=3532236 loops=1)
                           Sort Key: threadmemberships_1.userid, threads.channelid
                           Sort Method: external merge  Disk: 449368kB
                           ->  Hash Join  (cost=106322.79..282999.31 rows=3532236 width=120) (actual time=502.282..2055.334 rows=3532236 loops=1)
                                 Hash Cond: ((threadmemberships_1.postid)::text = (threads.postid)::text)
                                 ->  Seq Scan on threadmemberships threadmemberships_1  (cost=0.00..82419.36 rows=3532236 width=60) (actual time=0.009..416.028 rows=3532236 loops=1)
                                 ->  Hash  (cost=86640.35..86640.35 rows=846835 width=60) (actual time=502.091..502.092 rows=847646 loops=1)
                                       Buckets: 65536  Batches: 32  Memory Usage: 2897kB
                                       ->  Seq Scan on threads  (cost=0.00..86640.35 rows=846835 width=60) (actual time=0.445..352.307 rows=847646 loops=1)
                     ->  Sort  (cost=1072322.18..1084012.75 rows=4676228 width=60) (actual time=7251.193..8494.440 rows=4676227 loops=1)
                           Sort Key: channelmembers.userid, channelmembers.channelid
                           Sort Method: external merge  Disk: 320360kB
                           ->  Seq Scan on channelmembers  (cost=0.00..202637.28 rows=4676228 width=60) (actual time=8.958..808.813 rows=4676228 loops=1)
 Planning Time: 12.105 ms
 JIT:
   Functions: 33
   Options: Inlining true, Optimization true, Expressions true, Deforming true
   Timing: Generation 0.882 ms, Inlining 36.530 ms, Optimization 114.044 ms, Emission 83.245 ms, Total 234.701 ms
 Execution Time: 30914.343 ms
(27 rows)
```

The modified one has been running for almost three hours and has not finished:
```
mattermost=> select now();
              now
-------------------------------
 2023-03-31 23:36:18.742082+02
(1 row)

mattermost=> select * from pg_stat_activity where query like 'EXPLAIN%' \gx
-[ RECORD 1 ]----+------------------------------------------------------------------------------------------------------------
datid            | 16384
datname          | mattermost
pid              | 10627
usesysid         | 16385
usename          | mmuser
application_name | psql
client_addr      |
client_hostname  |
client_port      | -1
backend_start    | 2023-03-31 20:40:37.902438+02
xact_start       | 2023-03-31 20:41:57.57126+02
query_start      | 2023-03-31 20:42:07.774892+02
state_change     | 2023-03-31 20:42:07.774896+02
wait_event_type  |
wait_event       |
state            | active
backend_xid      | 1629
backend_xmin     | 1629
query            | EXPLAIN ANALYZE DELETE FROM threadmemberships WHERE (postid, userid) IN (                                  +
                 |     SELECT                                                                                                 +
                 |         threadmemberships.postid,                                                                          +
                 |         threadmemberships.userid                                                                           +
                 |     FROM                                                                                                   +
                 |         threadmemberships                                                                                  +
                 |         JOIN threads ON threads.postid = threadmemberships.postid                                          +
                 |     WHERE                                                                                                  +
                 |         (threadmemberships.UserId, threads.ChannelId) NOT IN (SELECT UserId, ChannelId FROM channelmembers)+
                 | );
backend_type     | client backend
```

Reply:
@agarciamontoro oh that's great! It would be worrying if we had so much difference in results.

BTW upon a discussion with @jespino he came up with another idea for a query to try

```sql
DELETE FROM threadmemberships AS tm USING (
    SELECT
        subtm.postid,
        subtm.userid
    FROM
        threadmemberships AS subtm
        JOIN threads ON threads.postid = subtm.postid
        LEFT JOIN channelmembers ON channelmembers.userid = subtm.userid
            AND threads.channelid = channelmembers.channelid
    WHERE
        channelmembers.channelid IS NULL) AS orphanmemberships
WHERE
    tm.postid = orphanmemberships.postid
    AND tm.userid = orphanmemberships.userid;
```

In which the explain looks like it's better, but the actual execution time is a bit worse.
Can you also try this version in your machine?

Reply:
FYI Alejandro is on PTO for the next 2 weeks